streams.go

```diff
@@ -0,0 +1,311 @@
+	err := e.publishStreamHeader(header, destinationIdentities)
+	if err != nil {
+		return nil, err
+	}
+
+	onWrite := func(data []byte) (int, error) {
+		return writeStreamBytes(data, e, info.Id, destinationIdentities, info.Size, onProgress)
+	}
+
+	onClose := func() error {
+		return e.publishStreamTrailer(info.Id, destinationIdentities)
+	}
+
+	return &ByteStreamWriter{
+		baseStreamWriter: newBaseStreamWriter[[]byte](onWrite, onClose),
+		Info:             info,
+	}, nil
+}
+
+type baseStreamReader struct {
+	readBuffer    bytes.Buffer
+	totalByteSize *uint64
+	bytesReceived int
+
+	closed     bool
+	closedLock sync.Mutex
+
+	onProgress *func(progress float64)
+}
+
+func newBaseStreamReader(totalByteSize *uint64) *baseStreamReader {
+	baseReader := &baseStreamReader{
+		bytesReceived: 0,
+	}
+	if totalByteSize != nil {
+		baseReader.totalByteSize = totalByteSize
+	}
+	return baseReader
+}
+
+func (r *baseStreamReader) enqueue(chunk *protocol.DataStream_Chunk) {
+	r.closedLock.Lock()
+	defer r.closedLock.Unlock()
+
+	if r.closed {
+		return
+	}
+	// write seems to handle growing the buffer if needed
+	r.readBuffer.Write(chunk.Content)
+}
+
+func (r *baseStreamReader) OnProgress(onProgress *func(progress float64)) {
+	r.onProgress = onProgress
+}
+
+func (r *baseStreamReader) maybeCallOnProgress(n int) {
+	r.bytesReceived += n
+
+	if r.totalByteSize != nil && r.onProgress != nil {
+		currentProgress := float64(r.bytesReceived) / float64(*r.totalByteSize)
+		(*r.onProgress)(currentProgress)
+	}
+}
+
+func (r *baseStreamReader) Read(bytes []byte) (int, error) {
+	n, err := r.readBuffer.Read(bytes)
+	r.maybeCallOnProgress(n)
+	return n, err
+}
+
+func (r *baseStreamReader) ReadByte() (byte, error) {
+	n, err := r.readBuffer.ReadByte()
+	r.maybeCallOnProgress(1)
+	return n, err
+}
+
+func (r *baseStreamReader) ReadBytes(delim byte) ([]byte, error) {
+	n, err := r.readBuffer.ReadBytes(delim)
+	r.maybeCallOnProgress(len(n))
+	return n, err
+}
+
+func (r *baseStreamReader) close() {
+	r.closedLock.Lock()
+	defer r.closedLock.Unlock()
+
+	if !r.closed {
+		r.closed = true
+	}
+}
+
+type TextStreamReader struct {
+	*baseStreamReader
+	Info TextStreamInfo
+}
+
+func NewTextStreamReader(info TextStreamInfo, totalChunkCount *uint64) *TextStreamReader {
+	return &TextStreamReader{
+		baseStreamReader: newBaseStreamReader(totalChunkCount),
+		Info:             info,
+	}
+}
+
+func (r *TextStreamReader) ReadRune() (rune, int, error) {
+	n, size, err := r.readBuffer.ReadRune()
+	r.maybeCallOnProgress(size)
+	return n, size, err
+}
+
+func (r *TextStreamReader) ReadString(delim byte) (string, error) {
+	n, err := r.readBuffer.ReadString(delim)
+	r.maybeCallOnProgress(len(n))
+	return n, err
+}
+
+func (r *TextStreamReader) ReadAll() string {
+	// wait for the stream to be closed
+	// not sure if this is the best way to do this
+	for {
+		r.closedLock.Lock()
+		if r.closed {
```

Comment:
can `closed` be atomic to avoid all the extra locking just for that one variable. Maybe, `readBuffer.Write` ☝️ requires a lock?

Reply:
Right yea that should be behind a lock